Fix link color for roles without unfiltered_html capabilities

oandregal opened the pull request:
Fixes https://github.com/WordPress/gutenberg/issues/25151

oandregal committed:
Tell kses not to filter out link color prop

oandregal committed:
Tell kses to allow the values we pass to link color

oandregal committed:
Make linter happy

oandregal committed:
Provide more linter happiness

nosolosw commented on lib/global-styles.php:
```diff
@@ -840,6 +840,45 @@ function gutenberg_experimental_global_styles_register_cpt() {
 add_action( 'init', 'gutenberg_experimental_global_styles_register_cpt' );
 add_filter( 'block_editor_settings', 'gutenberg_experimental_global_styles_settings' );
 add_action( 'wp_enqueue_scripts', 'gutenberg_experimental_global_styles_enqueue_assets' );
+add_filter( 'safe_style_css', 'gutenberg_experimental_global_styles_allow_css_var_name' );
```
kses has several checks when filtering post content, [this](https://github.com/WordPress/wordpress-develop/blob/master/src/wp-includes/kses.php#L2150) is the relevant function when it comes to filter the `style` attribute:

- check whether the property name is one of the allowed ([see](https://github.com/WordPress/wordpress-develop/blob/master/src/wp-includes/kses.php#L2177))
- check that the property value doesn't have extraneous characters, excluding parentheses as well ([see](https://github.com/WordPress/wordpress-develop/blob/master/src/wp-includes/kses.php#L2381))

nosolosw commented on lib/global-styles.php:
```diff
@@ -840,6 +840,45 @@ function gutenberg_experimental_global_styles_register_cpt() {
+	// The only value the block editor attaches to link color is
+	// var(--wp--preset--color--<some-value-here>)
+	// so be specific in testing for that value.
+	return preg_match( '/var\(--wp--preset--color--.*\)/', $parts[1] );
```
mmm, I believe we want to improve this bit. Allowing any characters (`.`) may still be problematic security-wise. It feels like we want to allow only alphanumeric and `-` as valid characters (eg of potential names: `...--some-color-value`, `...--some-color-value-2`, etc), which should cover all the potential color names that people use.

nosolosw replied:
improved at https://github.com/WordPress/gutenberg/pull/25411/commits/d27c8698c322aa007664861ac2ef37878104907b

oandregal committed:
Improve regular expression

- match start and end of value characters
- make sure color name can only have hyphen or alphanumeric chars

oandregal committed:
Add e2e test

oandregal committed:
prettify

jorgefilipecosta commented on lib/global-styles.php:
```diff
@@ -840,6 +840,45 @@ function gutenberg_experimental_global_styles_register_cpt() {
+function gutenberg_experimental_global_styles_allow_css_var_value( $allow_css, $css_test_string ) {
+	$parts = explode( ':', $css_test_string, 2 );
+
+	if ( '--wp--style--color--link' !== $parts[0] ) {
```
Should we apply trim before comparing to remove whitespace?

nosolosw replied:
https://github.com/WordPress/gutenberg/pull/25411/commits/7b94a13c896245a566c67e07a363ed201687f520

jorgefilipecosta commented on lib/global-styles.php:
```diff
@@ -840,6 +840,45 @@ function gutenberg_experimental_global_styles_register_cpt() {
+	// The only value the block editor attaches to link color is
+	// var(--wp--preset--color--<value-with-alphanumeric-chars-or-hyphen>)
+	// so be specific in testing for that value.
+	return preg_match( '/^var\(--wp--preset--color--[A-Za-z0-9-]*\)$/', $parts[1] );
```
I guess trim here may also be helpful.

nosolosw replied:
https://github.com/WordPress/gutenberg/pull/25411/commits/7b94a13c896245a566c67e07a363ed201687f520

jorgefilipecosta commented on lib/global-styles.php:
```diff
@@ -840,6 +840,45 @@ function gutenberg_experimental_global_styles_register_cpt() {
+	// The only value the block editor attaches to link color is
+	// var(--wp--preset--color--<value-with-alphanumeric-chars-or-hyphen>)
+	// so be specific in testing for that value.
+	return preg_match( '/^var\(--wp--preset--color--[A-Za-z0-9-]*\)$/', $parts[1] );
```
We are only accepting the variable values for the link color, which means that currently, custom link colors are not working.

I guess the custom color values are already accepted so we can fix the issue by simply return true if $allow_css is already true. Meaning we will not apply any additional restriction if an attribute is accepted we will continue to accept it but we will also accept the CSS variable references.

nosolosw replied:
Good catch, thanks!

nosolosw replied:
fixed at https://github.com/WordPress/gutenberg/pull/25411/commits/7b94a13c896245a566c67e07a363ed201687f520

jorgefilipecosta reviewed (commented):
In the core, we already have a good testing infrastructure for kses rule changes. And we will need to port this change to core anyways. Would it make sense the create a core patch with the change here and adding some kses rule test cases there? Similar to the patch we previously implemented for gradients https://github.com/WordPress/wordpress-develop/pull/110/files.

oandregal committed:
Pass through if value is a valid CSS color

nosolosw commented:
> In the core, we already have a good testing infrastructure for kses rule changes. And we will need to port this change to core anyways. Would it make sense the create a core patch with the change here and adding some kses rule test cases there?

I can do that as a follow-up to this. I'd also want to merge this as it is to make sure the link color is no longer broken in the plugin without having to wait for a core release that is still a few months away.

What I'm unsure about is: _when_ to prepare/merge that core patch. Is the 5.6 window a good time? I know there's uncertainty about whether link color is going to be merged into core in 5.6 (it doesn't look like it) so I wonder if we should wait to the 5.7 window instead. To my understanding, people have talked about that a few things can change (the general approach -classes instead of css vars-, the name of the var, the name of the value). I haven't done a core patch before so I don't have a good sense of whether this is a good idea or not (things like how difficult would it be to remove/update the patch, should we need it). So, essentially, I'm happy to take any advice here. What do you think?

oandregal committed:
Make linter happy

jorgefilipecosta reviewed (approved):
LGTM 👍
The reason I referred the core patch was to be able to test this kses changes against the core kses tests to increase the confidence we have on this change given that it may have a security impact. It would be good to test the patch before the next Gutenberg release.

nosolosw commented:
I'm trying to figure out who can provide a sanity check from a security point of view before this is merged.

nosolosw commented:
I've run this by some security folks and they aren't wild about serializing CSS Custom Properties in the post content. They'd rather keep the behavior as it is (not allowing CSS vars in the post content) or allow any in core, as it was done for `data-*` attributes in the past.

So we're back at the drawing board. Going to close this for the moment. Perhaps we can revisit Riad's https://github.com/WordPress/gutenberg/pull/21420
